hrjobcontract/js/src/job-contract/controllers/modal/modal-contract.controller.js

```diff
@@ -362,32 +306,202 @@ define([
+    }
+
+    /**
+     * Initializes the form entity fields by having a copy of the form entity
+     * and converting date strings into date objects.
+     */
+    function initFormEntity () {
+      angular.copy(entity, $scope.entity);
+      $scope.entity.details.period_start_date = convertToDateObject($scope.entity.details.period_start_date);
+      $scope.entity.details.period_end_date = convertToDateObject($scope.entity.details.period_end_date);
+    }
+
+    /**
+     * Handles all the confirmations that need to be displayed when updating the contract.
+     */
+    function processContractUpdate () {
+      if (angular.equals(entity, $scope.entity) &&
+        angular.equals(files, $scope.files) &&
+        !$scope.uploader.details.contract_file.queue.length &&
+        !$scope.uploader.pension.evidence_file.queue.length) {
+        $scope.$broadcast('hrjc-loader-hide');
+        $modalInstance.dismiss('cancel');
+        return;
+      }
+
+      switch (action) {
+        case 'edit':
+          if ($scope.entity.contract.is_primary === entity.contract.is_primary) {
+            confirmEdit().then(function (confirmed) {
+              switch (confirmed) {
+                case 'edit':
+                  contractEdit();
+                  break;
+                case 'change':
+                  changeReason().then(function (results) {
+                    contractChange(results.reasonId, results.date);
+                  });
+                  break;
+              }
+            });
+          } else {
+            contractEdit();
+          }
+          break;
+        case 'change':
+          changeReason().then(function (results) {
+            contractChange(results.reasonId, results.date);
+          });
+          break;
+        default:
+          $scope.$broadcast('hrjc-loader-hide');
+          $modalInstance.dismiss('cancel');
+      }
+    }
+
+    /**
+     * Runs the contract saving sequence which includes:
+     * - Validating the contract dates.
+     * - Confirming if entitlements should be updated.
+     * - Updating the contract.
+     */
+    function save () {
+      if (!$scope.allowSave) {
+        return;
+      }
+
+      $scope.$broadcast('hrjc-loader-show');
+      contractDetailsService.validateDates({
         contact_id: settings.contactId,
-        effective_date: date
+        period_start_date: $scope.entity.details.period_start_date,
+        period_end_date: $scope.entity.details.period_end_date,
+        jobcontract_id: entity.contract.id
       }).then(function (result) {
         if (result.success) {
-          saveContractChange(reasonId, date);
+          processContractUpdate();
         } else {
           CRM.alert(result.message, 'Error', 'error');
           $scope.$broadcast('hrjc-loader-hide');
         }
       }, function (reason) {});
+      $scope.$broadcast('hrjc-loader-hide');
     }
 
     function saveContractChange (reasonId, date) {
```

Comment:
It is a bit weird... We have functions 'save', 'contractEdit', 'saveContractChange'. Could you please provide a description for this func so at least we know the difference?